File: src/core/tracing.ts
import { randomBytes } from "node:crypto";
import { appendFile, mkdir } from "node:fs/promises";
import { dirname, resolve } from "node:path";

export type TraceAttribute = string | number | boolean;
export type TraceAttributes = Record<string, TraceAttribute>;
export type TraceStatusCode = "UNSET" | "OK" | "ERROR";

/** 接近 OpenTelemetry SpanData 的最小、稳定 JSON 结构。时间使用 Unix 纳秒字符串。 */
export type TraceSpanRecord = {
  traceId: string;
  spanId: string;
  parentSpanId?: string;
  name: string;
  kind: "INTERNAL" | "CLIENT";
  startTimeUnixNano: string;
  endTimeUnixNano: string;
  attributes: TraceAttributes;
  status: { code: TraceStatusCode; message?: string };
};

/** 生产环境可以实现这个小接口，把 span 转交给真正的 OTel SDK/exporter。 */
export type TraceExporter = {
  export(span: TraceSpanRecord): Promise<void> | void;
};

export type StartSpanOptions = {
  parent?: TraceSpan;
  kind?: TraceSpanRecord["kind"];
  attributes?: TraceAttributes;
};

/**
 * Tracer 只负责创建 span，不知道 Agent、模型或工具的具体实现。
 * exporter 失败默认不会打断 Agent 主流程，调用者可通过 onExportError 接入日志系统。
 */
export class AgentTracer {
  constructor(
    private readonly exporter: TraceExporter,
    private readonly onExportError: (error: unknown) => void = () => undefined,
    private readonly now: () => number = Date.now,
  ) {}

  startSpan(name: string, options: StartSpanOptions = {}): TraceSpan {
    return new TraceSpan(
      this,
      name,
      options.parent?.traceId ?? randomHex(16),
      randomHex(8),
      options.parent?.spanId,
      options.kind ?? "INTERNAL",
      options.attributes ?? {},
      unixNano(this.now()),
    );
  }

  async finish(record: TraceSpanRecord): Promise<void> {
    try {
      await this.exporter.export(record);
    } catch (error) {
      this.onExportError(error);
    }
  }

  timestamp(): string {
    return unixNano(this.now());
  }
}

export class TraceSpan {
  private ended = false;

  constructor(
    private readonly tracer: AgentTracer,
    readonly name: string,
    readonly traceId: string,
    readonly spanId: string,
    readonly parentSpanId: string | undefined,
    private readonly kind: TraceSpanRecord["kind"],
    private readonly initialAttributes: TraceAttributes,
    private readonly startTimeUnixNano: string,
  ) {}

  async end(
    code: TraceStatusCode = "OK",
    attributes: TraceAttributes = {},
    error?: unknown,
  ): Promise<void> {
    if (this.ended) return;
    this.ended = true;
    const message = error instanceof Error ? error.message : undefined;
    await this.tracer.finish({
      traceId: this.traceId,
      spanId: this.spanId,
      ...(this.parentSpanId ? { parentSpanId: this.parentSpanId } : {}),
      name: this.name,
      kind: this.kind,
      startTimeUnixNano: this.startTimeUnixNano,
      endTimeUnixNano: this.tracer.timestamp(),
      attributes: {
        ...this.initialAttributes,
        ...attributes,
        ...(error instanceof Error
          ? { "error.type": error.name, "error.message": error.message }
          : {}),
      },
      status: { code, ...(message ? { message } : {}) },
    });
  }
}

/** 一行一个 span，既适合教学查看，也方便 jq、Vector 或 Fluent Bit 采集。 */
export class JsonlTraceExporter implements TraceExporter {
  private pending: Promise<void> = Promise.resolve();
  private readonly filePath: string;

  constructor(filePath: string) {
    if (!filePath.trim()) throw new Error("Trace file path cannot be empty.");
    this.filePath = resolve(filePath);
  }

  export(span: TraceSpanRecord): Promise<void> {
    const line = `${JSON.stringify(span)}\n`;
    const writing = this.pending.then(async () => {
      await mkdir(dirname(this.filePath), { recursive: true });
      await appendFile(this.filePath, line, "utf8");
    });
    // 让单次写入错误返回给 Tracer 记录，但不要让队列永久停留在 rejected 状态。
    this.pending = writing.catch(() => undefined);
    return writing;
  }
}

export function createTracerFromEnvironment(
  environment: Readonly<Record<string, string | undefined>> = process.env,
  onExportError?: (error: unknown) => void,
): AgentTracer | undefined {
  const filePath = environment.AGENT_TRACE_FILE?.trim();
  if (!filePath) return undefined;
  return new AgentTracer(new JsonlTraceExporter(filePath), onExportError);
}

function randomHex(bytes: number): string {
  return randomBytes(bytes).toString("hex");
}

function unixNano(milliseconds: number): string {
  return (BigInt(Math.trunc(milliseconds)) * 1_000_000n).toString();
}

File: python/from_scratch_agent/tracing.py
"""零依赖、OpenTelemetry-compatible 的 Agent tracing。"""

import json
import secrets
import time
from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path
from threading import Lock
from typing import Protocol


@dataclass
class SpanRecord:
    trace_id: str
    span_id: str
    name: str
    kind: str
    start_time_unix_nano: str
    end_time_unix_nano: str
    attributes: dict[str, str | int | float | bool] = field(default_factory=dict)
    status: dict[str, str] = field(default_factory=dict)
    parent_span_id: str | None = None


class SpanExporter(Protocol):
    def export(self, span: SpanRecord) -> None:
        """导出一个已经结束的 span。"""


class JsonlTraceExporter:
    """一行写入一个 span；Lock 防止并行工具把两行内容交叉写坏。"""

    def __init__(self, file_path: str | Path) -> None:
        self.file_path = Path(file_path).resolve()
        self._lock = Lock()

    def export(self, span: SpanRecord) -> None:
        # 文件字段与 TypeScript 版保持一致，便于同一个采集器统一处理。
        payload = {
            "traceId": span.trace_id,
            "spanId": span.span_id,
            **(
                {"parentSpanId": span.parent_span_id}
                if span.parent_span_id
                else {}
            ),
            "name": span.name,
            "kind": span.kind,
            "startTimeUnixNano": span.start_time_unix_nano,
            "endTimeUnixNano": span.end_time_unix_nano,
            "attributes": span.attributes,
            "status": span.status,
        }
        line = json.dumps(payload, ensure_ascii=False) + "\n"
        with self._lock:
            self.file_path.parent.mkdir(parents=True, exist_ok=True)
            with self.file_path.open("a", encoding="utf-8") as file:
                file.write(line)


class Tracer:
    def __init__(
        self,
        exporter: SpanExporter,
        on_export_error: Callable[[Exception], None] = lambda _error: None,
    ) -> None:
        self.exporter = exporter
        self.on_export_error = on_export_error

    def start_span(
        self,
        name: str,
        parent: "Span | None" = None,
        kind: str = "INTERNAL",
        attributes: dict[str, str | int | float | bool] | None = None,
    ) -> "Span":
        return Span(
            tracer=self,
            name=name,
            trace_id=parent.trace_id if parent else secrets.token_hex(16),
            span_id=secrets.token_hex(8),
            parent_span_id=parent.span_id if parent else None,
            kind=kind,
            attributes=dict(attributes or {}),
            start_time_unix_nano=str(time.time_ns()),
        )

    def finish(self, span: SpanRecord) -> None:
        try:
            self.exporter.export(span)
        except Exception as error:
            # 观测系统故障不应该让模型或工具主流程失败。
            self.on_export_error(error)


@dataclass
class Span:
    tracer: Tracer
    name: str
    trace_id: str
    span_id: str
    parent_span_id: str | None
    kind: str
    attributes: dict[str, str | int | float | bool]
    start_time_unix_nano: str
    _ended: bool = False

    def end(
        self,
        code: str = "OK",
        attributes: dict[str, str | int | float | bool] | None = None,
        error: BaseException | None = None,
    ) -> None:
        if self._ended:
            return
        self._ended = True
        combined = {**self.attributes, **(attributes or {})}
        if error:
            combined["error.type"] = type(error).__name__
            combined["error.message"] = str(error)
        self.tracer.finish(
            SpanRecord(
                trace_id=self.trace_id,
                span_id=self.span_id,
                parent_span_id=self.parent_span_id,
                name=self.name,
                kind=self.kind,
                start_time_unix_nano=self.start_time_unix_nano,
                end_time_unix_nano=str(time.time_ns()),
                attributes=combined,
                status={
                    "code": code,
                    **({"message": str(error)} if error else {}),
                },
            )
        )
